File: 课程网站/microlectures/2.2-convolution/source/first_experiment.m
% Standalone full linear convolution. Keep time origins separately.
x=[1 2 3 4];h=[3 2 1];x0=0;h0=0;
nx=x0+(0:numel(x)-1);nh=h0+(0:numel(h)-1);
y=conv(x,h);ny=x0+h0+(0:numel(y)-1);manual=zeros(size(y));
for i=1:numel(y),for j=1:numel(x),k=ny(i)-nx(j)-h0+1;if k>=1 && k<=numel(h),manual(i)=manual(i)+x(j)*h(k);end;end;end
assert(max(abs(y-manual))<1e-12);disp('Rows: time index; output');disp([ny;y]);
f=figure('Visible','off');subplot(3,1,1);stem(nx,x);title('Input');subplot(3,1,2);stem(nh,h);title('Impulse response');subplot(3,1,3);stem(ny,y);title('Full linear convolution');xlabel('n (sample index)');
exportgraphics(f,fullfile(fileparts(mfilename('fullpath')),'first_experiment_matlab.png'));close(f);

File: 课程网站/microlectures/2.2-convolution/source/first_experiment.py
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
x=np.array([1,2,3,4]);h=np.array([3,2,1]);x0=0;h0=0
nx=x0+np.arange(len(x));nh=h0+np.arange(len(h));y=np.convolve(x,h);ny=x0+h0+np.arange(len(y))
manual=[sum(v*h[n-k-h0] for k,v in zip(nx,x) if 0<=n-k-h0<len(h)) for n in ny]
assert np.array_equal(manual,y);print('time:',ny,'output:',y)
fig,axes=plt.subplots(3,1,figsize=(9,7))
for ax,xs,ys,title in zip(axes,[nx,nh,ny],[x,h,y],['Input','Impulse response','Full linear convolution']):ax.stem(xs,ys);ax.set_title(title);ax.set_xlabel('n (sample index)')
fig.tight_layout();fig.savefig(Path(__file__).with_suffix('.png'));plt.close(fig)
